File: frontend/src/services/stock-ledger.ts
/**
 * Stock Ledger API Service
 * Immutable audit trail of all stock changes
 */

import { http } from './http';
import type { PaginatedResponse } from '@/types/common';

export type StockTransactionType = 'IN' | 'OUT' | 'ADJUST' | 'TRANSFER';

export interface StockLedgerEntry {
    id: number;
    product_id: number;
    warehouse_id: number;
    type: StockTransactionType;
    qty_change: number;
    qty_before: number;
    qty_after: number;
    reference_type?: string;
    reference_id?: number;
    note?: string;
    created_by?: number;
    created_at: string;
    product?: {
        id: number;
        sku: string;
        name: string;
    };
    warehouse?: {
        id: number;
        code: string;
        name: string;
    };
}

export interface StockLedgerFilter {
    page?: number;
    page_size?: number;
    product_id?: number;
    warehouse_id?: number;
    type?: StockTransactionType;
    date_from?: string;
    date_to?: string;
    reference_type?: string;
}

/**
 * Stock Ledger API endpoints
 */
export const stockLedgerApi = {
    async list(filters: StockLedgerFilter = {}) {
        const query = new URLSearchParams();
        if (filters.page !== undefined) query.append('page', String(filters.page));
        if (filters.page_size !== undefined) query.append('page_size', String(filters.page_size));
        if (filters.product_id !== undefined) query.append('product_id', String(filters.product_id));
        if (filters.warehouse_id !== undefined) query.append('warehouse_id', String(filters.warehouse_id));
        if (filters.type) query.append('type', filters.type);
        if (filters.date_from) query.append('date_from', filters.date_from);
        if (filters.date_to) query.append('date_to', filters.date_to);
        if (filters.reference_type) query.append('reference_type', filters.reference_type);

        const url = query.toString() ? `/stock-ledger?${query.toString()}` : '/stock-ledger';
        return http.get<PaginatedResponse<StockLedgerEntry>>(url);
    },

    async getById(id: number) {
        return http.get<StockLedgerEntry>(`/stock-ledger/${id}`);
    },

    /**
     * Get stock movement history for a product
     */
    async getProductHistory(productId: number, limit = 50) {
        return http.get<StockLedgerEntry[]>(`/stock-ledger?product_id=${productId}&page_size=${limit}`);
    },

    /**
     * Get stock movement history for a warehouse
     */
    async getWarehouseHistory(warehouseId: number, limit = 50) {
        return http.get<PaginatedResponse<StockLedgerEntry>>(`/stock-ledger?warehouse_id=${warehouseId}&page_size=${limit}`);
    },
};

File: backend/app/schemas/stock_ledger.py
"""StockLedger schemas."""

from pydantic import BaseModel, Field, field_serializer
from typing import Optional
from datetime import datetime

from app.models.stock_ledger import StockTransactionType


class StockLedgerBase(BaseModel):
    """Base stock ledger schema."""
    product_id: int
    warehouse_id: int
    type: StockTransactionType
    qty_change: int
    qty_before: int
    qty_after: int
    reference_type: Optional[str] = Field(None, max_length=50)
    reference_id: Optional[int] = None
    note: Optional[str] = None


class StockLedgerCreate(BaseModel):
    """
    Schema for creating a stock ledger entry.
    Used internally by apply_stock_change method.
    """
    product_id: int
    warehouse_id: int
    type: StockTransactionType
    qty_change: int
    qty_before: int
    qty_after: int
    reference_type: Optional[str] = None
    reference_id: Optional[int] = None
    note: Optional[str] = None
    created_by: Optional[int] = None


class StockLedgerResponse(StockLedgerBase):
    """Schema for stock ledger response."""
    id: int
    created_by: Optional[int] = None
    created_at: datetime

    class ProductSummary(BaseModel):
        id: int
        sku: str
        name: str

        class Config:
            from_attributes = True

    class WarehouseSummary(BaseModel):
        id: int
        code: str
        name: str

        class Config:
            from_attributes = True

    product: Optional[ProductSummary] = None
    warehouse: Optional[WarehouseSummary] = None

    @field_serializer("type")
    def serialize_type(self, value: StockTransactionType) -> str:
        """Return enum name (IN/OUT/ADJUST/TRANSFER) for frontend compatibility."""
        return value.name

    class Config:
        from_attributes = True


class StockLedgerFilter(BaseModel):
    """Schema for filtering stock ledger entries."""
    product_id: Optional[int] = None
    warehouse_id: Optional[int] = None
    type: Optional[StockTransactionType] = None
    reference_type: Optional[str] = None
    reference_id: Optional[int] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
